### ISSUE TITLE
twisted.python.rebuild broken under PyPy2

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9226]
|[<img alt="dansut's avatar" src="https://avatars.githubusercontent.com/u/100365?s=50" width="50" height="50">](https://github.com/dansut)| @dansut reported|
|-|-|
|Trac ID|trac#9226|
|Type|defect|
|Created|2017-07-21 22:47:52Z|

The `twisted.test.test_rebuild.RebuildTests.testFileRebuild` has a FAIL status on PyPy2 but that is not the half of it... other rebuild stuff also on top of it (`updateInstance`) can also potentially, seemingly randomly, fail under current PyPy.

I've traced the issue to problems with `gc.get_referrers(Cls)` not always bringing back everything it should... PyPy bug here logged here:  https://bitbucket.org/pypy/pypy/issues/2612/

This patch provides a workaround until PyPy fixed

```
diff --git a/src/twisted/python/rebuild.py b/src/twisted/python/rebuild.py
index 22cdecf7b..df7ca3603 100644
--- a/src/twisted/python/rebuild.py
+++ b/src/twisted/python/rebuild.py
@@ -210,12 +210,13 @@ def rebuild(module, doLog=1):
             clazz.__module__ = module.__name__
     if newclasses:
         import gc
+        gc.collect()
     for nclass in newclasses:
         ga = getattr(module, nclass.__name__)
         if ga is nclass:
             log.msg("WARNING: new-class %s not replaced by reload!" % reflect.qual(nclass))
         else:
-            for r in gc.get_referrers(nclass):
+            for r in gc.get_objects():
                 if getattr(r, '__class__', None) is nclass:
                     r.__class__ = ga
     if doLog:
```

`twisted.python.rebuild` doesn't even really work under PY3 yet (ticket:8887 ticket:8213) so PyPy3 is def a no go ...

There's also another related rebuild PyPy ticket:8575 that may or may not be fixed ... probably needs a test ...

<details><summary>Searchable metadata</summary>

```
trac-id__9226 9226
type__defect defect
reporter__dansut dansut
priority__normal normal
milestone__PyPy_support PyPy-support
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1500677272572277 1500677272572277
changetime__1512006307962490 1512006307962490
version__None None
owner__markrwilliams markrwilliams

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dansut's avatar" src="https://avatars.githubusercontent.com/u/100365?s=50" width="50" height="50">](https://github.com/dansut)<a name="note_2"></a>|@dansut set owner to @dansut|
|-|-|

Armin Rigo of PyPy worked out the exact problem here (see [PyPY ticket 2612](https://bitbucket.org/pypy/pypy/issues/2612/)) and has just landed [a commit which should fix](https://bitbucket.org/pypy/pypy/commits/3c4fb99e0c59) this.

However it's not going to be massively efficient so the suggestion is actually to implement in a similar (but more efficient) way to my original patch - have a PR to land once [ticket:8213]'s PR is merged.

**Comment by twisted-trac** [comment]
|[<img alt="dansut's avatar" src="https://avatars.githubusercontent.com/u/100365?s=50" width="50" height="50">](https://github.com/dansut)<a name="note_3"></a>|@dansut commented|
|-|-|

As of the release of PyPy 5.9.0 this problem does not show for both the 2.7 and 3.5 builds of PyPy.

twisted.test.test_rebuild.RebuildTests.testFileRebuild PASSES

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc set owner to @markrwilliams|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

Verified the fix with PyPy 5.9.0:

https://twistedmatrix.com/pipermail/twisted-python/2017-November/031754.html

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc set status to `closed`|
|-|-|

Verified the fix with PyPy 5.9.0:

https://twistedmatrix.com/pipermail/twisted-python/2017-November/031754.html
